django/db/models/expressions.py

```diff
@@ -485,12 +485,23 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        if self.value is None:
+        val = self.value
+        if self._output_field_or_none is not None:
+            if self.for_save:
```

Comment:
If the only use of for_save is in here, we could just have it as an attribute of the compiler. (Insert/UpdateCompiler has a class-level attribute for_save=True, other compilers have it as False)